### PR TITLE
Read from followers2: APM-296

### PR DESCRIPTION
**Pull request by neunhoef**
This is stage2 of Read from Followers.
This time it adds the forgotten outgoing HTTP-Header
  x-arango-potential-dirty-read: true
in all cases in which dirty reads could be handed out.

We need to do the following:
  - it is the transaction that decides about dirty reads, not the
    individual operation, therefore we have to ask the transaction, if
    the outgoing HTTP header has to be set
  - this has to be done in all RestHandlers, which can deliver dirty
    reads: RestDocumentHandler, RestCursorHandler, RestEdgesHandler,
    RestTransactionHandler
  - for queries, we need to extremely careful: at first, the query
    transaction decides, the query then copies the information about
    dirty reads for the time, when the query transaction is gone,
    QueryResults copy the information from the query, for the time when
    the Query object is gone, and finally, Cursors also copy the
    information, the QueryResultCursor keeps it in its QueryResult,
    the QueryStreamCursor keeps a copy (which it loads from the Query)
  - coordinator forwarding is not a problem, since the forwarding code
    basically forwards all HTTP headers on the way in and out

### Scope & Purpose

- [*] :hankey: Bugfix
- [*] :pizza: New feature

### Checklist

- [*] Tests
  - [*] **integration tests**
- [*] :book: CHANGELOG entry made
- [*] :books: documentation written (release notes, API changes, ...)
- [*] Backports: no backports

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [*] Docs PR: https://github.com/arangodb/docs/pull/1048, https://github.com/arangodb/arangodb/pull/16477
- [*] Enterprise PR: ???
- [*] GitHub issue / Jira ticket: APM-296
- [*] Design document: https://github.com/arangodb/documents/blob/master/DesignDocuments/02_PLANNING/ReadFromFollowers.md

### CODE REVIEW

File: arangod/Aql/Query.cpp [hunk 1]
```diff
@@ -384,6 +385,11 @@ std::unique_ptr<ExecutionPlan> Query::preparePlan() {
   _trx->addHint(
       transaction::Hints::Hint::FROM_TOPLEVEL_AQL);  // only used on toplevel
 
+  // We need to preserve the information aout dirty reads, since the
```
Comment (jsteemann) [inline arangod/Aql/Query.cpp:388]:
```suggestion
  // We need to preserve the information about dirty reads, since the
```

File: arangod/Aql/QueryCursor.h [hunk 2]
```diff
@@ -72,6 +74,11 @@ class QueryResultCursor final : public arangodb::Cursor {
   /// If no extras are set this will return a NONE slice.
   arangodb::velocypack::Slice extra() const;
 
+  /// @brief Remember, if we dirty reads were allowed:
```
Comment (jsteemann) [inline arangod/Aql/QueryCursor.h:77]:
```suggestion
  /// @brief Remember, if dirty reads were allowed:
```

File: arangod/Aql/QueryCursor.h [hunk 3]
```diff
@@ -24,9 +24,11 @@
 
 #pragma once
 
+#include "Aql/Query.h"
 #include "Aql/QueryResult.h"
 #include "Aql/SharedAqlItemBlockPtr.h"
 #include "Basics/Common.h"
+#include "StorageEngine/TransactionState.h"
```
Comment (jsteemann) [inline arangod/Aql/QueryCursor.h:31]: The 2 new includes don't seem to be needed by the new code added to this header. Can we move them to the .cpp file?

File: arangod/RestHandler/RestDocumentHandler.cpp [hunk 4]
```diff
@@ -386,6 +386,10 @@ RestStatus RestDocumentHandler::readSingleDocument(bool generateBody) {
     return RestStatus::DONE;
   }
 
+  if (_activeTrx->state()->options().allowDirtyReads) {
+    setOutgoingDirtyReadsHeader(true);
+  };
```
Comment (jsteemann) [inline arangod/RestHandler/RestDocumentHandler.cpp:391]:
```suggestion
  }
```

File: arangod/RestHandler/RestDocumentHandler.cpp [hunk 5]
```diff
@@ -854,6 +858,10 @@ RestStatus RestDocumentHandler::readManyDocuments() {
...
```
Comment (jsteemann) [inline arangod/RestHandler/RestDocumentHandler.cpp:863]:
```suggestion
  }
```

File: arangod/Aql/QueryResult.h [hunk 6]
```diff
@@ -90,6 +90,11 @@ struct QueryResult {
   std::shared_ptr<arangodb::velocypack::Builder> data;
   std::shared_ptr<arangodb::velocypack::Builder> extra;
   std::shared_ptr<transaction::Context> context;
+  bool allowDirtyReads;  // indicate that query was done with dirty reads,
```
[the remark this answers is not in the document]
Comment (neunhoef) [reply arangod/Aql/QueryResult.h:93]: Good catch, done!
